Add support for source and target tracks (to promote from/to)

kevin-david opened the pull request:
Note: I have not tested this yet, just wanted to get it out there for feedback first in case you'd like me to rework it.

<hr/>

When I release my app, I first move from `alpha` to `beta`, then `beta` to `production`.

These changes add support for arbitrary track names, so I can [hopefully!] use this action to achieve that flow.

The inputs are not required for backwards compatibility - as in, the old hardcoded `beta`->`production` promotion remains the default.

Commit:
Add support for source and target tracks (to promote from/to)

When I release my app, I first move from `alpha` to `beta`, then `beta` to `production`.

These changes add support for arbitrary track names, so I can use this action to achieve that flow.

kevin-david commented on src/promote-play-release.js:
```diff
@@ -11,6 +11,8 @@ export async function run() {
     const rawServiceAccountJson = getInput('service-account-json-raw', { required: true });
     const inAppUpdatePriorityInput = getInput('inapp-update-priority');
     const userFractionInput = getInput('user-fraction');
+    const fromTrack = getInput('from-track') ?? 'beta';
```
alternative names are `source-track` and `target-track` maybe? from/to felt more natural with the "promotion" concept though

boswelja replied:
Agreed, from/to makes more sense in this case

kevin-david commented on package-lock.json:
```diff
@@ -5,6 +5,7 @@
   "requires": true,
   "packages": {
     "": {
+      "name": "promote-play-beta-action",
```
Not sure why `npm i` did this... but seems safe :shrug:

kevin-david commented:
@boswelja would appreciate a review on this when you have a chance! i'm assuming you're OK with this enhancement generally but if not let me know :)

boswelja commented on action.yml:
```diff
@@ -18,6 +18,12 @@ inputs:
     description: "Portion of users who should get the staged version of the app. Accepts values between 0.0 and 1.0 (exclusive-exclusive)"
     required: false
+  from-track:
+    description: "The source track to promote from. If not supplied, `beta` is used."
```
I'm not super familiar with Actions conventions, but would it make more sense to give this a default value and set required to true, rather than hide the default in the action code?

kevin-david replied:
Good point - a default should be set: https://github.com/actions/toolkit/blob/main/packages/core/README.md#inputsoutputs, so I added it here instead.

I noticed the default is missing for `user-fraction` - should that be 1.0 or 0.0 in your mind? https://github.com/boswelja/AutoEvent/blob/main/.github/workflows/production-publish.yml#L15 doesn't have it... what does "undefined" do right now?

Commit:
Add defaults for from/to track in `action.yml`

Following the guidance here: https://github.com/actions/toolkit/blob/main/packages/core/README.md#inputsoutputs.

Also update README and description to match existing "default" conventions.

Commit:
Run the build

Commit:
Adding a missing await

Commit:
Resolve path to serviceAccountJson early

Commit:
Add default for user-fraction to action.yml

Commit:
Add missing publisher parameter

Commit:
Fixing async in storeServiceAccountJson

kevin-david commented:
Hey @boswelja I ended up having a few issues (you might be able to tell by the past few commits) and ended up taking this and porting it to typescript since the lack of types was bothering me :) 

I have it uploaded here: https://github.com/kevin-david/promote-play-release

If this is still something you're actively using & interested in, I can look into porting the changes back. Or, I can just add you as a maintainer on that new repo or something...

boswelja commented:
Yep, that's fine. Normally I'd say pulling the changes in to this repo would make more sense so we don't break anything for any other users, but dropping "beta" from the name makes more sense.
I'm still actively using the action, but I haven't had much time to work on it either. If you want to open a pull request to add a deprecation notice here pointing to your repo that would probably be ideal. I'll let you decide whether to add me as a maintainer to the new repo 🙂

kevin-david commented:
Closing this in favor of https://github.com/boswelja/promote-play-beta-action/pull/72
